planet/cli/data.py

```diff
@@ -0,0 +1,118 @@
+    """Turn filter JSON into a dict."""
+    return json.loads(value)
+
+
+@click.group()
+@click.pass_context
+def data(ctx):
+    """A group of commands for interacting with the Data API."""
+
+
+# TODO: filter().
+
+
+@data.command()
+# Our CLI command functions are async functions. They need an event
+# loop to be executed. The coro decorator provides that event loop.
+@coro
+# The command has two positional arguments. ITEM_TYPES comes first.
+@click.argument("item_types", callback=parse_item_types)
+@click.argument("filter", callback=parse_filter)
+# The command has three options.
+@click.option("--limit",
+              type=int,
+              default=100,
+              help="Maximum number of results to return.")
+@click.option("--name", help="Name for the saved search.")
+@click.option("--pretty", is_flag=True, help="Pretty print output.")
+# This decorator gives our function a context (named "ctx") that may
+# contain parameters set by the "planet data" command group, like
+# "planet data --quiet" or "planet data -vvv".
+@click.pass_context
+async def search_quick(ctx,
+                       item_types: List[str],
+                       filter: dict,
+                       limit: int,
+                       pretty: bool,
+                       name: str):
+    """Execute a structured item search and print results.
+
+    Results are represented as a sequence of GeoJSON features.
+
+    """
+
+    # Note: API tokens will be found in "ctx". They are unused in this
+    # example.
+    #
+    # The Planet Data client will return a stream of GeoJSON
+    # Feature-like dicts (aka an "iterator"). The search_quick function
+    # will iterate over the features, encode them as JSON, and echo
+    # them. We simulate that here.
+    async def example_client(item_types,
+                             filter,
+                             limit=0,
+                             name=None) -> AsyncIterator[dict]:
+        """This is a placeholder.
+
+        We'll delete this and use the real method from
+        planet.clients.data when it is available.
+
+        """
+        # Note: we re-emit the input item types and filter to help us
+        # sanity check during early sprints.
+        for feature in item_types:
+            yield dict(item_type=feature)
+
+        yield dict(filter=filter)
+
+    indent = 2 if pretty else None
+
+    # CLI functions raise ClickException when any kind of intentionally
+    # raised PlanetError occurs. We *don't* handle other errors now
+    # because those will be caused by bugs in our code and we want them
+    # raw and unfiltered.
+    try:
+        async for feature in example_client(item_types,
+                                            filter,
+                                            limit=limit,
+                                            name=name):
+            click.echo(json.dumps(feature, indent=indent))
+    except PlanetError as err:
+        raise click.ClickException() from err
```

Comment:
We'll use this pattern often. I got the idea of letting built-in errors propagate unhandled from Brett Slatkin. A ClickException raised from PlanetError tells a user that they are using the CLI wrong. A built-in exception like ValueError or TypeError can be a signal that we have a bug in our code.

Reply:
here I recommend using `@translate_exceptions` wrapper which handles exceptions with https://github.com/planetlabs/planet-client-python/blob/9fd238ba96bcd828e39b8cac3dda54efdc2f7e83/planet/cli/cmds.py#L41